python/ray/data/dataset.py

```diff
@@ -2693,7 +2693,7 @@ def write_datasource(
                 soft=False,
             )
 
-        if hasattr(datasource, "write"):
+        if "write" in datasource.__dict__:
```

Comment:
This doesn't look right. The `datasource.__dict__` will be an empty dict. You may want to check `type(datasource).__dict__`

Reply:
good catch